### PR TITLE
Add ability to disallow repeated keys in CBOR

### PR DESCRIPTION
**Pull request by timmc**
Fixes https://github.com/Kotlin/kotlinx.serialization/issues/2662 by adding a `visitKey` method to `CompositeDecoder`; map and set serializers should call this so that decoders have an opportunity to throw an error when a duplicate key is detected. Calls to this method are added to:

- `MapLikeSerializer.readElement` to handle maps
- `CborReader.decodeElementIndex` to handle data classes

A new config option `Cbor.forbidDuplicateKeys` can be set to true to enable this new behavior. This can form the basis of a Strict Mode in the future.

### CODE REVIEW

File: core/commonMain/src/kotlinx/serialization/SerializationExceptions.kt [hunk 1]
```diff
@@ -133,3 +133,10 @@ internal constructor(message: String?) : SerializationException(message) {
     // This constructor is used by the generated serializers
     constructor(index: Int) : this("An unknown field for index $index")
 }
+
+/**
+ * Thrown when a map deserializer encounters a repeated map key (and configuration disallows this.)
+ */
+@ExperimentalSerializationApi
+public class DuplicateMapKeyException(key: Any?) :
+    SerializationException("Duplicate keys not allowed in maps. Key appeared twice: $key")
```
Comment (timmc) [inline core/commonMain/src/kotlinx/serialization/SerializationExceptions.kt:142]: TODO: Just take a `message` param; I'm probably not going to end up doing assertions on key, and this is going to be annoying to generalize to sets, classes, etc.

File: core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt [hunk 2]
```diff
@@ -558,6 +558,18 @@ public interface CompositeDecoder {
...
+     * Called after a key has been read.
+     *
+     * This could be a map or set key, or anything otherwise intended to be
+     * distinct within the collection under normal circumstances.
+     *
+     * Implementations might use this as a hook for throwing an exception when
+     * duplicate keys are encountered.
+     */
+    @ExperimentalSerializationApi
+    public fun visitKey(key: Any?) { }
```
Comment (timmc) [inline core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:572]: I'm not happy with this name but couldn't think of anything better -- suggestions very welcome. :-)

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 3]
```diff
@@ -32,13 +32,14 @@ import kotlinx.serialization.modules.*
...
     internal val encodeDefaults: Boolean,
     internal val ignoreUnknownKeys: Boolean,
+    internal val allowDuplicateKeys: Boolean,
     override val serializersModule: SerializersModule
 ) : BinaryFormat {
 
     /**
      * The default instance of [Cbor]
      */
-    public companion object Default : Cbor(false, false, EmptySerializersModule())
```
Comment (timmc) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:41]: Is there a convention in this library that the defaults should be false, or is the introduction of a `true`-defaulting setting OK? (Wasn't sure whether to go with allow or forbid in the name.)

Comment (timmc) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:41]: Leaning towards switching this to `forbidDuplicateKeys`, actually.

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.assertFailsWithMessage
+import kotlinx.serialization.decodeFromByteArray
+import kotlinx.serialization.HexConverter
+import kotlinx.serialization.DuplicateMapKeyException
+import kotlin.test.Test
+
+class CborStrictModeTest {
```
Comment (timmc) [inline formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt:9]: TODO: Add a (passing) test for duplicate key detection when deserializing a data class. I have a test locally that fails -- I think I need to add ` visitKey` call to some inlined code somewhere, but I haven't found the appropriate callsite yet.

File: core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt (hunk 2 above)
Comment (pdvrieze) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:572]: @xiaozhikang0916 The challenge is that the behaviour needs to be configurable. There is no central configuration mechanism in the library, only formats and `SerialDescriptors`. There is also the need to do this in an ABI/API compatible way. Formats already need to do various things to play nice/support the various library features. This would just be another one. It also makes sense as various formats have their own duplication semantics.

Comment (timmc) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:572]: I initially tried plumbing the information through on the deserialization layer and it went very badly. This is only the start of what would be required: https://github.com/Kotlin/kotlinx.serialization/compare/dev...timmc:timmc/all-strict (although in practice it would be a config object, not a single boolean -- this was just a failed proof of concept.)

File: formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt (hunk 4 above)
Comment (timmc) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt:9]: I think I need to modify the serialization plugin's code generator in order to do this! Where does that code even live?

Comment (timmc) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt:9]: Hmm. I had previously tried modifying `decodeElementIndex` and wasn't getting the results I expected. Maybe I'll take another crack at it.

Comment (timmc) [reply formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborStrictModeTest.kt:9]: Oh! I see, the `decodeElementIndex` in `CborReader`. Got it.

File: core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt (hunk 2 above)
Comment (xiaozhikang0916) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:572]: I mean the duplication check in other formats ( if they need ), would share some common logic, like holding a set and putting keys in it. These common logic can be extracted to a utility class.

------

On the other hand, I am against adding this function in `Decoding` interface. It is implemented by each format with different behaviours, and only for internal using, which means not public, and users writing custom serializers should never call this function.

IMO, you may probably leave some instructions for format developers about how to achieve duplication check, rather than a public default funtion in this base interface, and make it open for users.

Comment (pdvrieze) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:572]: It would be possible to circumvent the MapSerializer if you have a map specific decoder. This decoder could intercept the decoding of the key, and then check there (before the result is returned to the serializer).

Comment (timmc) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:573]: I'm happy to implement whatever is needed, but I don't have a good understanding of what would be acceptable to the maintainers. It sounds like visitKey isn't well-liked, and the options might be:

- A. Each format that wants a duplicate-detection option provides its own replacement for a `MapLikeSerializer`-like serializer, fully reimplementing the logic. (This would be done via a serializers module, right?)
- B. Copy `MapLikeSerializer` to a new utility class `NoDuplicatesMapLikeSerializer` and add the duplicate detection; formats use this in their serializers module if desired.
- C. Like B, but just change `MapLikeSerializer` to make `readElement` non-final so that it can be overridden by classes like `HashMapSerializer` (or copies of them) rather than copying the implementation.
- D. Change the map *builder* to reject duplicate keys (haven't looked into feasibility here)

Which of these seems most palatable? Are there other possibilities?

Comment (timmc) [reply core/commonMain/src/kotlinx/serialization/encoding/Decoding.kt:573]: One concern I have with anything that involves the serializers module is that it risks interference with application-configured serializers. Each relevant format would have to document warnings like "if you specify an alternative serializer for maps, you'll need to reimplement duplicate key detection". Having it be performed in shared code (i.e. MapLikeSerializer) would at least allow the warning to recommend subclassing that shared code.
